### PR TITLE
feat: add recursive option for readdir

### LINKED ISSUE #197: fs.readdir recursive doesn't work
hello,

files = await fs.readdir(someDir,{recursive:true})

doesn't work. 
only the files in someDir are returned. 
it seems the options are ignored.

it would be nice if recrsion would work.

### PR DESCRIPTION
**Pull request by shulandmimi**
Issue fix #197

*Description of changes:*

add recursive option for readdir

```js
fs.readdir('example', { recursive: true })
    .then(res => {
        console.log(res);
    });

```

### CODE REVIEW

File: src/fs/read_dir.rs [hunk 1]
```diff
@@ -57,25 +64,57 @@ pub async fn read_dir<'js>(
...
+            }
+        }
+    };
     let mut items = Vec::with_capacity(64);
+    let mut dirs = VecDeque::from_iter([PathBuf::from(path)]);
 
-    while let Some(child) = dir.next_entry().await? {
-        if let Some(name) = child.path().file_name() {
-            let name = name.to_string_lossy().to_string();
+    while let Some(mut dir_buf) = dirs.pop_back() {
```
Comment (richarddavison) [inline src/fs/read_dir.rs:100]: Nice we have a non-recursive traversal 👍 
I have similar functionality in here that could be changed to return an iterator instead and reused here.
https://github.com/awslabs/llrt/blob/37eb92b062f858a42567c1457ee199d977c51153/src/utils/io.rs#L45
Also, we might speed up things if we parallelize the walk of subdirectories by spawning tasks that joining the awaits that calls next entries.

File: fixtures/fs/readdir/readdir.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import fs from 'fs/promises';
+
+fs.readdir('./', { recursive: true }).then(res => {
+    console.log(res);
```
Comment (richarddavison) [inline fixtures/fs/readdir/readdir.js:4]: Do we need the console.log here?

File: fixtures/fs/readdir/recursive/readdir.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import fs from 'fs/promises';
+
+fs.readdir('./', { recursive: true }).then(res => {
+    console.log(res);
```
Comment (richarddavison) [inline fixtures/fs/readdir/recursive/readdir.js:4]: Same as above

File: src/fs/read_dir.rs [hunk 4]
```diff
@@ -57,25 +64,57 @@ pub async fn read_dir<'js>(
     path: String,
     options: Opt<Object<'js>>,
 ) -> Result<ReadDir> {
+    let mut path = path;
     let with_file_types = options
```
Comment (richarddavison) [inline src/fs/read_dir.rs:68]: When we now have two options. We can first check if we have options and then access the variables:
```rust
if let Some(opts) = opts.0{
 //check for withFileTypes & recursive option
}
```

**Review by richarddavison — CHANGES_REQUESTED** [review]
Thanks for the PR! Looks great, some minor adjustments! :)

File: src/fs/read_dir.rs (hunk 1 above)
Comment (shulandmimi) [reply src/fs/read_dir.rs:100]: Thanks for your suggestion, I added `DirectoryWalker` to replace them, you can see if they meet your expectations

**Review by richarddavison — APPROVED** [review]

**Review by richarddavison — APPROVED** [review]

**Pull request merged**